### PR TITLE
fix: handle case sensitivity for SQL catalog for snowflake

### PR DESCRIPTION
**Pull request by owlas**
Closes: #11134

see comments inline.

This was initially from an issue that said “pre-filled snowflake queries don’t work” but the root problem was worse and actually the schema browser and field browsers weren’t working correctly for snowflake users

### Before

CleanShot 2024-08-30 at 09 59 00@2x

### After
CleanShot 2024-08-30 at 14 39 50@2x

### CODE REVIEW

File: packages/backend/src/controllers/sqlRunnerController.ts [hunk 1]
```diff
@@ -70,21 +70,26 @@ export class SqlRunnerController extends BaseController {
         unauthorisedInDemo,
     ])
     @SuccessResponse('200', 'Success')
-    @Get('/tables/{tableName}')
+    @Get('/fields')
```
Comment (owlas) [inline packages/backend/src/controllers/sqlRunnerController.ts:73]: since this returns fields and not tables

File: packages/backend/src/controllers/sqlRunnerController.ts [hunk 2]
```diff
@@ -70,21 +70,26 @@ export class SqlRunnerController extends BaseController {
...
     @OperationId('getTableFields')
     async getTableFields(
         @Path() projectUuid: string,
-        @Path() tableName: string,
         @Request() req: express.Request,
-        @Query() schema?: string,
+        @Query() tableName?: string,
```
Comment (owlas) [inline packages/backend/src/controllers/sqlRunnerController.ts:78]: these are the filters for the fields - eventually it should accept database too

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 3]
```diff
@@ -3079,8 +3077,8 @@ export class ProjectService extends BaseService {
     async getWarehouseFields(
         user: SessionUser,
         projectUuid: string,
-        tableName: string,
-        schema?: string,
+        tableName?: string,
```
Comment (owlas) [inline packages/backend/src/services/ProjectService/ProjectService.ts:3080]: reflects that the API endpoint takes these as query params now so could be missing

File: packages/common/src/types/warehouse.ts [hunk 4]
```diff
@@ -83,6 +83,7 @@ export interface WarehouseClient {
     getFields(
         tableName: string,
         schema?: string,
+        database?: string,
```
Comment (owlas) [inline packages/common/src/types/warehouse.ts:86]: Allow database filtering in future

File: packages/frontend/src/features/sqlRunner/hooks/requestUtils.ts [hunk 5]
```diff
@@ -49,7 +49,10 @@ export const getResultsFromStream = async <T>(url: string | undefined) => {
         const result = await responseStream.text();
 
         // Split the JSON strings by newline
-        const jsonStrings = result.trim().split('\n');
+        const jsonStrings = result
```
Comment (owlas) [inline packages/frontend/src/features/sqlRunner/hooks/requestUtils.ts:52]: Unrelated fix when result returns 0 rows

File: packages/frontend/src/features/sqlRunner/hooks/useTableFields.tsx [hunk 6]
```diff
@@ -18,14 +18,18 @@ const fetchTableFields = async ({
     projectUuid,
     tableName,
     schema,
-}: Pick<GetTableFieldsParams, 'projectUuid' | 'tableName' | 'schema'>) =>
-    lightdashApi<WarehouseTableSchema>({
-        url: `/projects/${projectUuid}/sqlRunner/tables/${tableName}${
-            schema ? `?schema=${schema}` : ''
-        }`,
+}: Pick<GetTableFieldsParams, 'projectUuid' | 'tableName' | 'schema'>) => {
```
Comment (owlas) [inline packages/frontend/src/features/sqlRunner/hooks/useTableFields.tsx:21]: change to match new API

File: packages/warehouses/src/warehouseClients/BigqueryWarehouseClient.ts [hunk 7]
```diff
@@ -319,9 +319,10 @@ export class BigqueryWarehouseClient extends WarehouseBaseClient<CreateBigqueryC
     async getFields(
         tableName: string,
         schema: string,
+        database?: string,
     ): Promise<WarehouseCatalog> {
         const client = new BigQuery({
-            projectId: this.credentials.project,
+            projectId: database || this.credentials.project,
```
Comment (owlas) [inline packages/warehouses/src/warehouseClients/BigqueryWarehouseClient.ts:325]: bigquery uses the provided database or the default

File: packages/warehouses/src/warehouseClients/SnowflakeWarehouseClient.ts [hunk 8]
```diff
@@ -462,9 +462,9 @@ export class SnowflakeWarehouseClient extends WarehouseBaseClient<CreateSnowflak
         const whereSql = databaseName ? `AND TABLE_CATALOG ILIKE ?` : '';
         const query = `
             SELECT
-                LOWER(TABLE_CATALOG) as "table_catalog",
-                LOWER(TABLE_SCHEMA) as "table_schema",
-                LOWER(TABLE_NAME) as "table_name"
+                TABLE_CATALOG as "table_catalog",
```
Comment (owlas) [inline packages/warehouses/src/warehouseClients/SnowflakeWarehouseClient.ts:465]: This was the root cause of the bug - in the rest of Lightdash we force lowercase identifiers. But for the SQL Runner this is causing errors. Now all case is handled as case sensitive (see screenshots in PR)

**Review by rephus — COMMENTED** [review]
Code looks good, there are a couple of errors on build, so I can't test it on render. 

image

**Comment by owlas** [comment]
stupid tests

**Comment by owlas** [comment]
fixed!

**Comment by rephus** [comment]
Charts seem a bit broken , perhaps our functions don't support uppercase fields.

image

**Review by rephus — APPROVED** [review]
This PR fixes the issue from the ticket, but charts are not working as expected. I'll let you decide if this should be fixed here or on a separate PR

**Comment by rephus** [comment]
We should also update the API integration tests

image

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 9]
```diff
@@ -2073,30 +2075,27 @@ export class ProjectService extends BaseService {
...
         if (groupByColumns && groupByColumns.length > 0) {
             // Wrap the original query in a CTE
             let pivotedSql = `WITH original_query AS (${userSql}), group_by_query AS (${groupByQuery}), pivot_query AS (${pivotQuery})`;
 
-            pivotedSql += `\nSELECT * FROM pivot_query WHERE row_index < ${
+            pivotedSql += `\nSELECT * FROM pivot_query WHERE ${q}row_index${q} < ${
                 limit ?? 500
-            } and column_index < 10 order by row_index, column_index`;
-
+            } and ${q}column_index${q} < 10 order by ${q}row_index${q}, ${q}column_index${q}`;
```
Comment (joaoviana) [inline packages/backend/src/services/ProjectService/ProjectService.ts:2098]: completely forgot, why did we choose `10`?

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 10]
```diff
@@ -3002,9 +3006,7 @@ export class ProjectService extends BaseService {
             credentials,
         );
 
-        const schema = ProjectService.getWarehouseSchema(credentials);
-
-        const warehouseTables = await warehouseClient.getAllTables(schema);
+        const warehouseTables = await warehouseClient.getAllTables();
```
[the remark this answers is not in the document]
Comment (owlas) [reply packages/backend/src/services/ProjectService/ProjectService.ts:3009]: We get all tables for all schemas, the SQL Runner is supposed to show everything available to the user in the warehouse

File: packages/backend/src/services/ProjectService/ProjectService.ts (hunk 9 above)
Comment (owlas) [reply packages/backend/src/services/ProjectService/ProjectService.ts:2098]: It was just a default until we make it configurable!

**Review by joaoviana — APPROVED** [review]
